### PR TITLE
[WIP] Copy files from container

### PR DESCRIPTION
**Pull request by eaba**

### CODE REVIEW

File: src/DotNet.Testcontainers/Clients/TestcontainersClient.cs [hunk 1]
```diff
@@ -222,5 +222,10 @@ public Task<string> BuildAsync(IImageFromDockerfileConfiguration configuration,
     {
       return this.images.BuildAsync(configuration, ct);
     }
+
+    public Task ExportFilesFromContainerAsync(string id, string containerPath, Stream outStream, CancellationToken ct = default)
```
Comment (eaba) [inline src/DotNet.Testcontainers/Clients/TestcontainersClient.cs:226]: Here is how it was extracted https://github.com/GodelTech/CodeReview.Orchestrator/blob/c070b8960089a183a2bcaa0b3726b5a39c6e9c65/src/CodeReview.Orchestrator/Services/TarArchiveService.cs#L39-L69

File: src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs [hunk 2]
```diff
@@ -92,6 +92,12 @@ public Task ExtractArchiveToContainerAsync(string id, string path, Stream tarStr
       this.logger.ExtractArchiveToDockerContainer(id, path);
       return this.Docker.Containers.ExtractArchiveToContainerAsync(id, new ContainerPathStatParameters { Path = path, AllowOverwriteDirWithFile = false }, tarStream, ct);
     }
+    public async Task GetArchiveFromContainerAsync(string id, string path, Stream outStream, CancellationToken ct = default)
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs:95]: Add line break

File: src/DotNet.Testcontainers/Clients/IDockerContainerOperations.cs [hunk 3]
```diff
@@ -19,6 +19,7 @@ internal interface IDockerContainerOperations : IHasListOperations<ContainerList
     Task RemoveAsync(string id, CancellationToken ct = default);
 
     Task ExtractArchiveToContainerAsync(string id, string path, Stream tarStream, CancellationToken ct = default);
+    Task GetArchiveFromContainerAsync(string id, string path, Stream outStream, CancellationToken ct = default);
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/IDockerContainerOperations.cs:22]: Add line break

File: src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs [hunk 4]
```diff
@@ -92,6 +92,12 @@ public Task ExtractArchiveToContainerAsync(string id, string path, Stream tarStr
...
+    {
+      this.logger.GetArchiveFromContainer(id, path);
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/DockerContainerOperations.cs:97]:
```csharp
this.logger.GetArchiveFromContainer(id, path);

var archive = await this.Docker.Containers.GetArchiveFromContainerAsync(id, new GetArchiveFromContainerParameters { Path = path }, true, ct)
  .ConfigureAwait(false);

await archive.Stream.CopyToAsync(outStream, ct)
  .ConfigureAwait(false);
```

File: src/DotNet.Testcontainers/Clients/ITestcontainersClient.cs [hunk 5]
```diff
@@ -95,6 +96,18 @@ internal interface ITestcontainersClient
     /// <returns>Task that completes when the file has been copied.</returns>
     Task CopyFileAsync(string id, string filePath, byte[] fileContent, int accessMode, int userId, int groupId, CancellationToken ct = default);
 
+    /// <summary>
+    /// Copy file from container
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/ITestcontainersClient.cs:100]: `Copies a file from the container.`

File: src/DotNet.Testcontainers/Clients/ITestcontainersClient.cs [hunk 6]
```diff
@@ -95,6 +96,18 @@ internal interface ITestcontainersClient
...
+    /// </summary>
+    /// <param name="id">Docker container id.</param>
+    /// <param name="containerPath">Path to the file in the container</param>
+    /// <param name="outStream">Output stream</param>
+    /// <returns>Task that completes when the file has been copied.</returns>
+    ///
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/ITestcontainersClient.cs:106]: Fix comment and line breaks below (incl. L109, L110). `ct` has no param tag.

File: src/DotNet.Testcontainers/Clients/TestcontainersClient.cs (hunk 1 above)
Comment (HofmeisterAn) [reply src/DotNet.Testcontainers/Clients/TestcontainersClient.cs:226]: Don't you need the host path as well? I guess the part that extracts the archive is missing, right (thats your link)?

Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Clients/TestcontainersClient.cs:226]: `/// <inheritdoc />` missing

File: src/DotNet.Testcontainers/Containers/TestcontainersContainer.cs [hunk 7]
```diff
@@ -181,6 +182,15 @@ public Task CopyFileAsync(string filePath, byte[] fileContent, int accessMode =
...
 
+    public Task ExportFilesFromContainerAsync(string containerPath, Stream outStream)
+    {
+      if (string.IsNullOrWhiteSpace(containerPath))
+      {
+        throw new ArgumentException("Value cannot be null or whitespace.", nameof(containerPath));
+      }
+
+      return this.client.ExportFilesFromContainerAsync(this.Id, containerPath, outStream);
+    }
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Containers/TestcontainersContainer.cs:193]: Add line break

File: src/DotNet.Testcontainers/Containers/TestcontainersContainer.cs [hunk 8]
```diff
@@ -181,6 +182,15 @@ public Task CopyFileAsync(string filePath, byte[] fileContent, int accessMode =
       return this.client.CopyFileAsync(this.Id, filePath, fileContent, accessMode, userId, groupId, ct);
     }
 
+    public Task ExportFilesFromContainerAsync(string containerPath, Stream outStream)
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Containers/TestcontainersContainer.cs:185]: This should be part of the interface `IRunningDockerContainer` incl. documentation, etc. I suggest a different name too, relative to `CopyFileAsync`. Maybe `ReadFileAsync` or `GetFileAsync`.

File: src/DotNet.Testcontainers/Containers/TestcontainersContainer.cs (hunk 7 above)
Comment (HofmeisterAn) [reply src/DotNet.Testcontainers/Containers/TestcontainersContainer.cs:193]: Use guard. Something like:

```csharp
Guard.Argument(containerPath, nameof(containerPath))
  .NotNull()
  .NotEmpty();
```

File: src/DotNet.Testcontainers/Logging.cs [hunk 9]
```diff
@@ -93,6 +93,10 @@ public static void ExtractArchiveToDockerContainer(this ILogger logger, string i
     {
       _ExtractArchiveToDockerContainer(logger, path, id, null);
     }
+    public static void GetArchiveFromContainer(this ILogger logger, string id, string path)
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Logging.cs:96]: Add line break

File: src/DotNet.Testcontainers/Logging.cs [hunk 10]
```diff
@@ -93,6 +93,10 @@ public static void ExtractArchiveToDockerContainer(this ILogger logger, string i
...
+    {
+      _ExtractArchiveToDockerContainer(logger, path, id, null);
```
Comment (HofmeisterAn) [inline src/DotNet.Testcontainers/Logging.cs:98]: This log message does not fit.

**Review by HofmeisterAn — CHANGES_REQUESTED** [review]
Please cover the changes with unit tests, too. E.g. read a file from the container, copy it to the host and check the length, hash, etc. Also consider the SonarCloud Quality Gate.

**Comment by HofmeisterAn** [comment]
I'll close this pull request because it has not had recent activity. Don't hesitate to reopen it again. Maybe I find some time in the next days to start working on it.

**Pull request closed without merging**